File: lib/dotnet/AlphaSkia/AlphaSkiaTypeFace.cs
namespace AlphaSkia;

/// <summary>
/// Represents a typeface to draw text.
/// </summary>
public sealed class AlphaSkiaTypeFace : AlphaSkiaNative
{
    private readonly AlphaSkiaData? _nativeData;

    private AlphaSkiaTypeFace(IntPtr native, AlphaSkiaData? nativeData = null)
        : base(native, NativeMethods.alphaskia_typeface_free)
    {
        _nativeData = nativeData;
    }

    protected override void Dispose(bool disposing)
    {
        base.Dispose(disposing);
        if (disposing)
        {
            _nativeData?.Dispose();
        }
    }

    public static AlphaSkiaTypeFace? Register(byte[] data)
    {
        var nativeData = NativeMethods.alphaskia_data_new_copy(data, (ulong)data.LongLength);
        var typeface = NativeMethods.alphaskia_typeface_register(nativeData);
        if (typeface == IntPtr.Zero)
        {
            NativeMethods.alphaskia_data_free(nativeData);
            return null;
        }

        return new AlphaSkiaTypeFace(typeface, new AlphaSkiaData(nativeData));
    }

    public static AlphaSkiaTypeFace? Create(string name, bool bold, bool italic)
    {
        var typeface =
            NativeMethods.alphaskia_typeface_make_from_name(name, bold ? (byte)1 : (byte)0, italic ? (byte)1 : (byte)0);
        if (typeface == IntPtr.Zero)
        {
            return null;
        }

        return new AlphaSkiaTypeFace(typeface);
    }
}

File: lib/java/main/src/main/java/alphaskia/AlphaSkiaTypeface.java
package alphaskia;

public class AlphaSkiaTypeface extends AlphaSkiaNative {
    private final AlphaSkiaData data;

    private AlphaSkiaTypeface(long handle, AlphaSkiaData data) {
        super(handle);
        this.data = data;
    }

    @Override
    public void close() {
        AlphaSkiaData data = this.data;
        if (data != null) {
            data.close();
        }
        release(handle);
        handle = 0;
    }

    private native void release(long handle);

    public static AlphaSkiaTypeface register(byte[] data) {
        var nativeData = new AlphaSkiaData(data);
        var typeface = register(nativeData.handle);
        if (typeface == 0) {
            nativeData.close();
            return null;
        }
        return new AlphaSkiaTypeface(typeface, nativeData);
    }

    private static native long register(long handle);

    public static AlphaSkiaTypeface create(String name, boolean bold, boolean italic) {
        var typeface = makeFromName(name, bold, italic);
        if (typeface == 0) {
            return null;
        }

        return new AlphaSkiaTypeface(typeface, null);
    }

    private static native long makeFromName(String name, boolean bold, boolean italic);
}
